### DISCUSSION TITLE
Unable to create a View (Collection) for getting data from multiple collections

### DISCUSSION THREAD

**Discussion by pranjaldugar** [pocketbase/pocketbase#1938] [Q&A]
Hi folks at PocketBase,
I am trying to create a View with the following query
```sql
SELECT "", id, projects.donefor as person, projects.cost as total FROM projects WHERE projects.status = 'Payment Pending'
UNION
SELECT "", id, subscriptions.person as person, subscriptions.cost as total FROM subscriptions WHERE subscriptions.status = 'Payment Pending'
```
It is throwing an error `Invalid query - invalid identifier parts [projects.cost as total ""].`
however when I tried to modify the view using 'DB Browser for SQLite' and execute the following SQL
```sql
DROP VIEW "main"."payment_due";
CREATE VIEW `payment_due` AS SELECT * FROM (
SELECT "", id, projects.donefor as person, projects.cost as total FROM projects WHERE projects.status = 'Payment Pending'
UNION
SELECT "", id, subscriptions.person as person, subscriptions.cost as total FROM subscriptions WHERE subscriptions.status = 'Payment Pending'
)
```
Browse Data tab shows the correct result,
Not sure what I am missing over here? How can I get it working?

Thanks for support and great work!

**Comment by ganigeorgiev** [comment]
The parsing doesn't support `UNION` at the moment. I've added it to the [roadmap](https://github.com/orgs/pocketbase/projects/2) to consider it but for now is a low priority.

You can try to wrap the union expression in a `FROM`/subquery, eg.:

```sql
SELECT id, person, total
FROM (
    SELECT id, projects.donefor as person, projects.cost as total FROM projects WHERE projects.status = 'Payment Pending'
    UNION
    SELECT id, subscriptions.person as person, subscriptions.cost as total FROM subscriptions WHERE subscriptions.status = 'Payment Pending'
)
```

Also please note that the first column with the empty string is not valid field name and it will fail the field validations. If you want to keep it, you can add a named alias, eg `"" as something`.

**Reply by pranjaldugar** [reply]
`"" as something`, ya it is not required,
your code does work, but a problem now I have with it is that it return the id of person as string, where as the person field is a relation
![image](https://user-images.githubusercontent.com/64412152/221807116-119d4e73-2814-4c98-bf11-3b6f6b7fdecd.png)

**Reply by ganigeorgiev** [reply]
It cannot be detected as a `relation` because the column value is constructed from 2 different fields (`projects.donefor` and `subscriptions.person`) where each could have different collection references.

Could you elaborate a little more on your use case and what are you trying to do?

**Reply by pranjaldugar** [reply]
donefor and person both relates to same collection 'persons'. 
tried changinig projects.donefor as projects.person (by changing name of column) still same issue persits.

**Reply by ganigeorgiev** [reply]
The parser and fields detection could be definitely improved but in the general sense when having union, subqueries or any other more complex expression, the result collection field would be ambiguous and even if they point to the same collection, their field options could be different - for example one of the field could be a single relation, while the other multiple, both could have different nonempty/unique constraints, etc.

I'm not sure that I understand your use case. Could you provide a little more context what are you trying to do, what is the purpose of the collections and why they hold identical information?

**Reply by pranjaldugar** [reply]
Sure, here is synops of my use case,
I have a firm where I sell some products, render some recurring servicess and also undertake custom projects,
I have kept them in 3 different collections as there contents are completely different.
for eg:  `projects` contains fields
![image](https://user-images.githubusercontent.com/64412152/221816045-e44df5d5-f988-4929-abc4-6e20a0bd66c7.png)
and `subscriptions` to reccuring servicess contains fields
![image](https://user-images.githubusercontent.com/64412152/221816425-eef5e44c-c0cc-44ec-9b49-4f61a432edf0.png)

and I have a collection `persons` referring to clients and freelancers.
`projects.donefor, projects.doneby, subscriptions.person` all are refering to 'persons' collection.

I am tring to create a view for finding how much payment is due to be collected from my clients/customes for projects they ordered and servicess they subscribed to.

I can do they following by writing it in js, but was thinking if it was possible to do it in UI itself as sql would be faster and I would not have to open other app just to get a view. I also don't have problem waiting for the parser being improved to handel relation is subqueries or UNION being supported but it could also help other use cases involving processing of multiple collections.

**Reply by ganigeorgiev** [reply]
I'll try in the near future to update the parser to resolve `UNION` queries but the behavior would be still ambiguous and it may not work as expected for everyone and I'll have to think a little more on it.

Take for example the following query:

```sql
SELECT id, "anything" as rel from A  -- static string literal as "rel" column
UNION
SELECT id, singleRel1 as rel from B  -- single relation field as "rel" column
UNION
SELECT id, singleRel2 as rel from C  -- single relation field as "rel" column (pointing to different collection)
UNION
SELECT id, multiRel as rel from D    -- multiple relation field as "rel" column
```

What should be the result field type of the `rel` column in this case?

We resolve unknown/ambigious fields to `json` because it is the safest field resolution since the `json` field type doesn't apply any normalizations on the value and will be returned as it is. And the above is just a simple example. In real scenario the query could be far more complicated and hard to predict to what field type it will yield.

---

If you want to process multiple collections and to preserve the field types, an alternative approach depending on your use case could be also to use `JOIN`(s) instead of `UNION`.

**Reply by ganigeorgiev** [reply]
@Anatoly03 You are commenting to a year old discussion that I no longer remember what actually was being discussed or have any information about your setup and collections structure to make sense of the above error.

Please don't hijack old closed posts and consider opening a new Q&A discussion with more details about your use case and the specific problem.
